File: codes/go/gpt/minimal/task_4/Solution.go
package main

import (
	"encoding/json"
	"fmt"
	"os"
)

func allocateResources(patients [][]int, resources []map[string]interface{}, responders int, totalTime int) map[string]interface{} {
	assignments := []map[string]int{}
	survivors := 0
	totalSurvivalProbability := 0

	for i := 0; i < len(patients); i++ {
		for j := 0; j < len(resources); j++ {
			resourceTime := int(resources[j]["time"].(float64))
			boost := int(resources[j]["boost"].(float64))
			if resourceTime <= totalTime {
				survivalProbability := patients[i][resourceTime-1] + boost
				if survivalProbability > 100 {
					survivalProbability = 100
				}
				if survivalProbability > 0 && responders > 0 {
					assignments = append(assignments, map[string]int{"patient": i, "resource": j, "time": resourceTime - 1})
					survivors++
					totalSurvivalProbability += survivalProbability
					responders--
					break
				}
			}
		}
	}

	return map[string]interface{}{
		"assignments":               assignments,
		"survivors":                 survivors,
		"total_survival_probability": totalSurvivalProbability,
	}
}

func main() {
	var input struct {
		Patients   [][]int               `json:"patients"`
		Resources  []map[string]interface{} `json:"resources"`
		Responders int                  `json:"responders"`
		TotalTime  int                  `json:"total_time"`
	}

	json.NewDecoder(os.Stdin).Decode(&input)

	result := allocateResources(input.Patients, input.Resources, input.Responders, input.TotalTime)

	output, _ := json.Marshal(result)
	fmt.Println(string(output))
}

File: codes/javascript/gpt/minimal/task_4/Solution.js
function allocateResources(patients, resources, responders, totalTime) {
    let n = patients.length;
    let assignments = [];
    let maxSurvivors = 0;
    let maxSurvivalProbability = 0;

    function calculateSurvivalProbability(patientIndex, resource, time) {
        let baseProbability = patients[patientIndex][time];
        let boostedProbability = baseProbability + resource.boost;
        return Math.min(boostedProbability, 100);
    }

    let dp = Array.from({ length: totalTime + 1 }, () => Array.from({ length: responders + 1 }, () => ({ survivors: 0, survivalProbability: 0, path: [] })));

    for (let t = 0; t <= totalTime; t++) {
        for (let r = 1; r <= responders; r++) {
            for (let patientIndex = 0; patientIndex < n; patientIndex++) {
                for (let resourceIndex = 0; resourceIndex < resources.length; resourceIndex++) {
                    let resource = resources[resourceIndex];
                    if (t + resource.time <= totalTime) {
                        let nextTime = t + resource.time;
                        let survivalProbability = calculateSurvivalProbability(patientIndex, resource, t);
                        let potentialSurvivors = dp[t][r - 1].survivors + (survivalProbability > 50 ? 1 : 0);
                        let potentialSurvivalProbability = dp[t][r - 1].survivalProbability + survivalProbability;

                        if (potentialSurvivors > dp[nextTime][r].survivors || (potentialSurvivors === dp[nextTime][r].survivors && potentialSurvivalProbability > dp[nextTime][r].survivalProbability)) {
                            dp[nextTime][r].survivors = potentialSurvivors;
                            dp[nextTime][r].survivalProbability = potentialSurvivalProbability;
                            dp[nextTime][r].path = [...dp[t][r - 1].path, { patient: patientIndex, resource: resourceIndex, time: t }];
                        }
                    }
                }
            }
        }
    }

    for (let r = 1; r <= responders; r++) {
        if (dp[totalTime][r].survivors > maxSurvivors || (dp[totalTime][r].survivors === maxSurvivors && dp[totalTime][r].survivalProbability > maxSurvivalProbability)) {
            maxSurvivors = dp[totalTime][r].survivors;
            maxSurvivalProbability = dp[totalTime][r].survivalProbability;
            assignments = dp[totalTime][r].path;
        }
    }

    return {
        assignments: assignments,
        survivors: maxSurvivors,
        total_survival_probability: maxSurvivalProbability
    };
}

function main() {
    const fs = require('fs');
    const input = JSON.parse(fs.readFileSync(0, 'utf-8'));
    const result = allocateResources(input.patients, input.resources, input.responders, input.total_time);
    console.log(JSON.stringify(result));
}

main();